clients/drcachesim/tracer/syscall_pt_trace.cpp

```diff
@@ -36,65 +36,85 @@
 
 /* For SYS_exit,SYS_exit_group. */
 #include "../../../core/unix/include/syscall_linux_x86.h"
-#include "dr_api.h"
-#include "drpttracer.h"
-#include "kcore_copy.h"
 #include "syscall_pt_trace.h"
 
 #ifndef BUILD_PT_TRACER
 #    error "This module requires the drpttracer extension."
 #endif
 
 #define PT_DATA_FILE_NAME_SUFFIX ".pt"
-#define PT_METADATA_FILE_NAME_SUFFIX ".metadata"
 
 syscall_pt_trace_t::syscall_pt_trace_t()
     : open_file_func_(nullptr)
     , write_file_func_(nullptr)
     , close_file_func_(nullptr)
     , pttracer_handle_ { GLOBAL_DCONTEXT, nullptr }
+    , pttracer_output_buffer_ { GLOBAL_DCONTEXT, nullptr }
     , recorded_syscall_count_(0)
     , cur_recording_sysnum_(-1)
     , drcontext_(nullptr)
-    , pt_dir_name_ { '\0' }
+    , output_file_(INVALID_FILE)
 {
 }
 
 syscall_pt_trace_t::~syscall_pt_trace_t()
 {
+    if (output_file_ != INVALID_FILE) {
+        close_file_func_(output_file_);
+        output_file_ = INVALID_FILE;
+    }
 }
 
 bool
-syscall_pt_trace_t::init(void *drcontext, char *pt_dir_name, size_t pt_dir_name_size,
+syscall_pt_trace_t::init(void *drcontext, char *pt_dir_name,
                          drmemtrace_open_file_func_t open_file_func,
                          drmemtrace_write_file_func_t write_file_func,
                          drmemtrace_close_file_func_t close_file_func)
 {
     drcontext_ = drcontext;
-    memcpy(pt_dir_name_, pt_dir_name, pt_dir_name_size);
     open_file_func_ = open_file_func;
     write_file_func_ = write_file_func;
     close_file_func_ = close_file_func;
     pttracer_handle_ = { drcontext, nullptr };
-    return true;
-}
-
-bool
-syscall_pt_trace_t::start_syscall_pt_trace(int sysnum)
-{
+    pttracer_output_buffer_ = { drcontext_, nullptr };
+    std::string output_file_name(pt_dir_name);
+    output_file_name +=
+        "/" + std::to_string(dr_get_thread_id(drcontext_)) + PT_DATA_FILE_NAME_SUFFIX;
+    output_file_ = open_file_func_(output_file_name.c_str(), DR_FILE_WRITE_REQUIRE_NEW);
 #define RING_BUFFER_SIZE_SHIFT 8
-    ASSERT(drcontext_ != nullptr, "drcontext_ is nullptr");
-    ASSERT(pttracer_handle_.handle == nullptr, "pttracer_handle_.handle isn't nullptr");
 
-    /* TODO i#5505: To reduce the overhead caused by pttracer initialization, we should
-     * share the same pttracer handle for all syscalls. We will do this when we upgrade
-     * drpt2ir to support decoding PT fragments that don't have PSB packets.
+    /* To reduce the overhead caused by pttracer initialization, we shared the same
```

Comment:
Add to comment more details about:
- how we may need to steal more FDs from the app with a large thread count
- how is an open handle treated when the thread migrates to a different CPU
- out of the two overheads (creating output buffer, opening PT handle), was one the dominant one? If yes, document that finding here.